File: python/dsa/algorithms/dp/2d/tsp.py
def solve_tsp(dist):
    n = len(dist)
    dp = [[float('inf')] * n for _ in range(1 << n)] # @viz: init
    
    dp[1][0] = 0 # @viz: base
    
    for mask in range(1, 1 << n): # @viz: loopMask
        for u in range(n): # @viz: loopU
            if not (mask & (1 << u)): continue
            
            for v in range(n): # @viz: loopV
                if mask & (1 << v): continue
                
                next_mask = mask | (1 << v) # @viz: next
                cost = dp[mask][u] + dist[u][v] # @viz: compute
                if cost < dp[next_mask][v]: # @viz: check
                    dp[next_mask][v] = cost # @viz: update
                    
    min_cost = float('inf') # @viz: resultInit
    for i in range(1, n): # @viz: resultLoop
        total = dp[(1 << n) - 1][i] + dist[i][0] # @viz: resultCompute
        if total < min_cost: # @viz: resultCheck
            min_cost = total # @viz: resultUpdate
            
    return min_cost # @viz: result

File: go/dsa/algorithms/dp/2d/tsp.go
func solveTSP(dist [][]int) int {
    n := len(dist)
    dp := make([][]int, 1 << n) // @viz: init
    for i := range dp { 
        dp[i] = make([]int, n)
        for j := range dp[i] { dp[i][j] = 1e9 }
    }
    
    dp[1][0] = 0 // @viz: base
    
    for mask := 1; mask < (1 << n); mask++ { // @viz: loopMask
        for u := 0; u < n; u++ { // @viz: loopU
            if mask & (1 << u) == 0 { continue }
            
            for v := 0; v < n; v++ { // @viz: loopV
                if mask & (1 << v) != 0 { continue }
                
                nextMask := mask | (1 << v) // @viz: next
                cost := dp[mask][u] + dist[u][v] // @viz: compute
                if cost < dp[nextMask][v] { // @viz: check
                    dp[nextMask][v] = cost // @viz: update
                }
            }
        }
    }
    
    minCost := 1000000000 // @viz: resultInit
    for i := 1; i < n; i++ { // @viz: resultLoop
        total := dp[(1 << n) - 1][i] + dist[i][0] // @viz: resultCompute
        if total < minCost { // @viz: resultCheck
            minCost = total // @viz: resultUpdate
        }
    }
    return minCost // @viz: result
}